### PR TITLE
Organisation Auth Service

### PR DESCRIPTION
**Pull request by dharmverma**

### CODE REVIEW

**Review by dpatel017 — APPROVED** [review]

File: Services/CO.CDP.Organisation.Authority/Endpoint.cs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    private static Model.TokenResponse CreateToken(string issuer, RsaSecurityKey rsaPrivateKey, string urn)
+    {
+        // TODO: Fetch additional claims for authorisation purpose
+
+        var tokenExpiry = 3600d;
+        var tokenDescriptor = new SecurityTokenDescriptor
+        {
+            Subject = new ClaimsIdentity(new[] {
+                    new Claim("sub", urn!),
```
Comment (jakzal) [inline Services/CO.CDP.Organisation.Authority/Endpoint.cs:80]: Why `!`?

File: Services/CO.CDP.Organisation.Authority/Endpoint.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                    return Results.BadRequest("Invalid client credentials");
+                }
+
+                return Results.Ok(CreateToken(issuer, rsaPrivateKey, urn!));
+            })
+            .DisableAntiforgery()
+            .WithMetadata(new ConsumesAttribute("application/x-www-form-urlencoded"))
+            .Produces<Model.TokenResponse>(StatusCodes.Status200OK, "application/json")
+            .Produces<ProblemDetails>(StatusCodes.Status500InternalServerError)
+            .Produces(StatusCodes.Status415UnsupportedMediaType); ;
```
Comment (jakzal) [inline Services/CO.CDP.Organisation.Authority/Endpoint.cs:69]: Bonus points for documenting the API 👍

File: TestKit/CO.CDP.TestKit.Mvc/TestWebApplicationFactory.cs [hunk 3]
```diff
@@ -9,6 +15,31 @@ public class TestWebApplicationFactory<TProgram>(Action<IHostBuilder> configurat
...
+        builder.ConfigureServices(services =>
+        {
+            services.AddSingleton<IPolicyEvaluator, FakePolicyEvaluator>();
+        });
+
         return base.CreateHost(builder);
     }
+}
+
+public class FakePolicyEvaluator : IPolicyEvaluator
```
Comment (jakzal) [inline TestKit/CO.CDP.TestKit.Mvc/TestWebApplicationFactory.cs:28]: ❤️

**Review by jakzal — APPROVED** [review]

**Review by glgoaco — APPROVED** [review]

File: Services/CO.CDP.Organisation.Authority.Tests/ApiTest.cs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    public ApiTest()
+    {
+        TestWebApplicationFactory<Program> factory = new(builder =>
+        {
+            builder.ConfigureAppConfiguration((context, config) =>
+            {
+                config.AddInMemoryCollection(
+                    [
+                        new ("Issuer",  _issuer),
+                        new ("PublicKey", "-----BEGIN RSA PUBLIC KEY-----\nMIIBCgKCAQEA0ND+OiUI9B/q2EARaXJzb8HaBUduUtkvxj7Qyb+5yw8JZ2/TO3aA0La6km+TLkBla9hwG8+NWTpx2tq+42YQU3L2274GI7dfUS8hqpJ6141PmchaO/10\ndII5e8GtRnaHqY8vWz5I03nPM9gYAT5h7bAlltN0scqCUnmvRmu1A0E5y65OT63o\nQIUAo9PpVqckxM0yHgYKlf6mtTu3dkd8LuhciRnNGaJ5/OGEeyeZrzBzgvE2zmrH\n3H66Oq9dkvr5K/Unhd7Lo4EdGyaI31nRSRshY7F/wVSIgsebya9tGHalh1pgSzL0\nqr9v1+WU5gG/XKvpFuIeByhSWvjsZJykiwIDAQAB\n-----END RSA PUBLIC KEY-----"),
```
[the remark this answers is not in the document]
Comment (dharmverma) [reply Services/CO.CDP.Organisation.Authority.Tests/ApiTest.cs:34]: Btw this is not token, that is public key

File: Libraries/CO.CDP.Swashbuckle/Security/SecurityExtensions.cs [hunk 5]
```diff
@@ -6,21 +6,21 @@ namespace CO.CDP.Swashbuckle.Security;
...
         {
-            Description = "GOV.UK One Login JWT Bearer token",
+            Description = "Organisation Authority JWT Bearer token",
             Type = SecuritySchemeType.Http,
             Scheme = "Bearer",
             Reference = new OpenApiReference
             {
                 Type = ReferenceType.SecurityScheme,
-                Id = "OneLogin"
+                Id = "OrganisationAuthority"
```
[the remark this answers is not in the document]
Comment (dharmverma) [reply Libraries/CO.CDP.Swashbuckle/Security/SecurityExtensions.cs:19]: Can you explain in more detail?

File: Frontend/CO.CDP.OrganisationApp/appsettings.json [hunk 6]
```diff
@@ -6,10 +6,12 @@
     }
   },
   "AllowedHosts": "*",
+  // https://docs.sign-in.service.gov.uk/integrate-with-integration-environment/managing-your-users-sessions
```
[the remark this answers is not in the document]
Comment (dharmverma) [reply Frontend/CO.CDP.OrganisationApp/appsettings.json:9]: This is for documentation purpose that why we have chosen 60 mins

File: Frontend/CO.CDP.OrganisationApp/ApiBearerTokenHandler.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            return payload!.AccessToken;
+        }
+        else
+        {
+            throw new Exception($"Unable to get access token from Organisation Authority, " +
+                $"{response.StatusCode} {await response.Content.ReadAsStringAsync()}");
+        }
+    }
+
+    public class TokenResponse
```
[the remark this answers is not in the document]
Comment (dharmverma) [reply Frontend/CO.CDP.OrganisationApp/ApiBearerTokenHandler.cs:68]: This is a required class for deserialization purpose. Model/TokenResponse.cs is in Api project, not available in Frontend App

File: Frontend/CO.CDP.OrganisationApp/ApiBearerTokenHandler.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+        if (userDetails != null)
+        {
+            if (userDetails.AccessToken == null)
+            {
+                var context = httpContextAccessor.HttpContext;
+                if (context != null)
+                {
+                    var oneLogintoken = await context.GetTokenAsync("access_token");
+                    // var expiresAt = DateTimeOffset.Parse(await context.GetTokenAsync("expires_at")).ToLocalTime();
```
[the remark this answers is not in the document]
Comment (dharmverma) [reply Frontend/CO.CDP.OrganisationApp/ApiBearerTokenHandler.cs:26]: This is here for debugging Onelogin token expiry time, which is helpful when things goes wonky.

**Pull request merged**